spec/docker_commander_spec.rb

```diff
@@ -62,4 +62,30 @@
     end
   end
 
+  describe '#handle_thin_pool_exception' do
+    it 'Raises when exception message is not related to thin pool' do
+      expect(Fastlane::Actions).to receive(:sh).with('docker pull bla').and_raise(FastlaneCore::Interface::FastlaneShellError, 'some message')
+      expect {
+        @docker_commander.handle_thin_pool_exception do
+          @docker_commander.pull_image(docker_image_name: 'bla')
+        end
+      }.to raise_error(FastlaneCore::Interface::FastlaneShellError, 'some message')
+    end
+
+    it 'Retries the command when the message is related to thin pool' do
+      expect(Fastlane::Actions).to receive(:sh).twice.with('docker pull bla').and_raise(FastlaneCore::Interface::FastlaneShellError, 'Create more free space in thin pool or ...')
+        @docker_commander.handle_thin_pool_exception do
+          @docker_commander.pull_image(docker_image_name: 'bla')
```

Comment:
Maybe, for clarity, you should directly call the method that will fail inside the block or you could raise inside the block:
```ruby
describe '#handle_thin_pool_exception' do
  it 'Raises when exception message is not related to thin pool' do
    expect {
      @docker_commander.handle_thin_pool_exception do
        raise FastlaneCore::Interface::FastlaneShellError, 'some message'
      end
    }.to raise_error(FastlaneCore::Interface::FastlaneShellError, 'some message')
  end
end
```
Then one would not be distracted by the mocking.